File: Array/Python/Permutation in String.py
# Problem link: https://leetcode.com/problems/permutation-in-string/description
# Time: O(26 * m + n)
# Space: O(1)
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        storeOne = [0] * 26
        storeTwo = [0] * 26

        for char in s1:
            storeOne[ord(char) - ord('a')] += 1
        
        for idx, char in enumerate(s2):
            storeTwo[ord(char) - ord('a')] += 1

            if idx >= len(s1):
                storeTwo[ord(s2[idx - len(s1)]) - ord('a')] -= 1
            
            if storeOne == storeTwo:
                return True
        
        return False

File: Array/Java/Permutation in String.java
// Problem link: https://leetcode.com/problems/permutation-in-string/description
// Time: O(26 * m + n)
// Space: O(1)
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int[] storeOne = new int[26];
        int[] storeTwo = new int[26];
        int sizeOne = s1.length(), sizeTwo = s2.length();

        for(char chr : s1.toCharArray()) {
            storeOne[chr - 'a']++;
        }

        for(int idx = 0; idx < sizeTwo; idx++) {
            storeTwo[s2.charAt(idx) - 'a']++;

            if(idx >= sizeOne) {
                storeTwo[s2.charAt(idx - sizeOne) - 'a']--;
            }

            if(Arrays.equals(storeOne, storeTwo)) {
                return true;
            }
        }

        return false;
    }
}
